### ISSUE TITLE
training stages

### ISSUE THREAD

**Issue by xjturobocon** [karfly/learnable-triangulation-pytorch#4]
Hey,
In the paper, when it comes to training, "The networks are trained for 6 epochs with 10−4 learning rate for the 2D backbone and a separate learning rate 10−3 for the volumetric backbone". I want to know if the the whole network is trained end-to-end after 2 seperate training?

**Comment by karfly** [comment]
Hey,

Actually, the whole training is don in end-to-end manner. We just set different learning rates for different parts of the network. That's how you can do it in PyTorch: https://pytorch.org/docs/stable/optim.html#per-parameter-options.

**Comment by xjturobocon** [comment]
Thanks you for your reply.
I mentioned that you did not apply any augmentations during the training, other than rotating the orientation of the cube in volumetric triangulation around the vertical axis. So is the rotation is important to the result?  The degree' range is [0, 360] or [0, 90]? or others? usually, in the test, the degree is set to zero, and I think the different degrees should not affect the result, but I found it did affect the result. different degrees resulted in different accuracy.  Can you kindly explain your implement?

**Comment by karfly** [comment]
During training we randomly rotate the cube around vertical axis in range [0, 360] degrees. During test we set this angle to zero.

**Comment by yurymalkov** [comment]
@xjturobocon We considered altering the cube angle in test (specifically, as a test-time augmentation), but didn't have time to evaluate it.

**Comment by xjturobocon** [comment]
Thanks.
Your loss function includes 2 terms, the second one is like − β · log( · ),  which maximizes the
prediction for the voxel that has inside of it the ground-truth joint.  I think in the initial stage, the number in the bracket can be minus, so the log function is meaningless, how did you solve it?

**Comment by yurymalkov** [comment]
@xjturobocon 
There is a space-wide softmax before the logarithm, so the input is always positive.

**Comment by xjturobocon** [comment]
> 
> 
> @xjturobocon
> There is a space-wide softmax before the logarithm, so the input is always positive.

**Comment by xjturobocon** [comment]
@yurymalkov 
I understand. I think the symbol in the paper may be mistakenly used, it represents the output before softmax. 

Another question is that have you tried to generate 3D heatmap as the supervision, and calculate 3D heatmaps loss? I found it hard for the network to learn in this way, but I don't konw why.

**Comment by yurymalkov** [comment]
@xjturobocon, thank for the comment!

Not sure, by heatmap loss you mean the L2 loss between gt and predicted heatmaps?

**Comment by xjturobocon** [comment]
yes.



---Original---

**Comment by yurymalkov** [comment]
L2 of MSE between heatmaps should work much worse compared to getting the center of max, as the loss does not directly optimizes the objective that we want. If you just extract the maximum it is ok (e.g. HRNet uses MSE and just extracts the maximum with simple post-processing, it works good on full-body, not so good on faces where accuracy could be very high). 
But it becomes very important when the quant of the space is bigger than the desired error (exactly our case the quant is ~4cm for 64x64x64 cube, the **average** L2 error can be as small as 1cm for CMU), so the loss should be exactly the objective that we want.

**Comment by xjturobocon** [comment]
I mean using L2 of MSE between heatmaps, and getting the center of max to calculate accuracy( not do softmax). However, the result is bad.

**Comment by yurymalkov** [comment]
@xjturobocon 
You can imagine a network trained with MSE between heatmaps loss that does not work well with center of max in test.

For instance, if you add small noise to prediction of each pixel MSE would stay almost the same, while the center of mass can be siginificantly shifted to the center of the image (due to multiplication of the prediction in a pixel by pixel's coordianates in the image, which can be ~100).

They are somewhat different objectives.

**Comment by xjturobocon** [comment]
I agree with that they are somewhat different objectives, but there are many 2D and 3D  huamn pose estimation works using MSE heatmap loss and getting center of max in test. (Coarse-to-fine human pose estimation network also used 3D volumetric representation and L2 loss of 3D heatmaps). They perform well. 

I found in my experiment that its performance is very poor. I think it is similar with coarse-to-fine work, and should get a reasonable result.

**Comment by yurymalkov** [comment]
@xjturobocon I agree, coarse-to-fine methods are likely to avoid this problem, as they are good at overcoming the limit of space quantization.
